// fsweb/src/parser/parse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function validate_port(input) {
    var numeric = +input;
    if (numeric === NaN)
        return false;
    var test = parseInt(input);
    return (test > 0 && test < 65536);
}
function validate_path(input) {
    return typeof input === "string";
}
function extract_input(argv) {
    var process = argv.shift();
    var script = argv.shift();
    var line = argv.join(' ');
    return line;
}
function extract_path(input) {
    var split = input.split(" ").map(function (n) { return n.trim(); }).filter(function (n) { return n.length > 0; });
    var path = undefined;
    if (split.length === 0) {
        path = "./";
    }
    else if (split.length >= 1) {
        if (validate_path(split[0])) {
            path = split[0];
        }
        else {
            throw Error("invalid path");
        }
    }
    return path;
}
function extract_port(input) {
    var split = input.split(" ").map(function (n) { return n.trim(); }).filter(function (n) { return n.length > 0; });
    var port = undefined;
    if (split.length === 0) {
        port = "5000";
    }
    else if (split.length === 1) {
        if (validate_port(split[0])) {
            port = split[0];
        }
        else {
            port = "5000";
        }
    }
    else if (split.length >= 2) {
        if (validate_port(split[1])) {
            port = split[1];
        }
        else {
            throw Error("invalid port");
        }
    }
    return parseInt(port);
}
function parse_argument(argv) {
    var input = extract_input(argv);
    return {
        path: extract_path(input),
        port: extract_port(input),
        timeout: 100
    };
}
exports.parse_argument = parse_argument;
